batch/batch/front_end/operators.py

```diff
@@ -0,0 +1,99 @@
+import abc
+from typing import List, Optional, Tuple
+
+
+class Operator(abc.ABC):
+    user_symbols: List[str]
+    negate: bool
+
+    @staticmethod
+    def is_match(s) -> bool:
+        return s in Operator.user_symbols
+
+    @staticmethod
+    def to_sql() -> str:
+        raise NotImplementedError
+
+
+class ComparisonOperator(Operator, abc.ABC):
+    user_symbols: List[str]
+    negate = False
+
+    def to_sql(self):
+        assert len(self.user_symbols) == 1
+        return self.user_symbols[0]
+
+
+class MatchOperator(Operator, abc.ABC):
+    pass
+
+
+class PartialMatchOperator(MatchOperator, abc.ABC):
+    def to_sql(self):
+        if self.negate:
+            return 'NOT LIKE'
+        return 'LIKE'
+
+
+class ExactMatchOperator(MatchOperator, abc.ABC):
+    def to_sql(self):
+        if self.negate:
+            return '!='
+        return '='
+
+
+class GreaterThanEqualOperator(ComparisonOperator):
+    user_symbols = ['>=']
+
+
+class LessThanEqualOperator(ComparisonOperator):
+    user_symbols = ['<=']
+
+
+class GreaterThanOperator(ComparisonOperator):
+    user_symbols = ['>']
+
+
+class LessThanOperator(ComparisonOperator):
+    user_symbols = ['<']
+
+
+class NotEqualExactMatchOperator(ExactMatchOperator, ComparisonOperator):
+    user_symbols = ['!=']
+    negate = True
+
+
+class EqualExactMatchOperator(ExactMatchOperator, ComparisonOperator):
+    user_symbols = ['==', '=']
+    negate = False
+
+
+class NotEqualPartialMatchOperator(PartialMatchOperator):
+    user_symbols = ['!~']
+    negate = True
+
+
+class EqualPartialMatchOperator(PartialMatchOperator):
+    user_symbols = ['=~']
+    negate = False
+
+
+ordered_operators = [
+    GreaterThanEqualOperator(),
+    LessThanEqualOperator(),
+    GreaterThanOperator(),
+    LessThanOperator(),
+    NotEqualPartialMatchOperator(),
+    EqualPartialMatchOperator(),
+    NotEqualExactMatchOperator(),
+    EqualExactMatchOperator(),
+]
+
+
+def parse_operator(term) -> Optional[Tuple[Operator, str, str]]:
+    for op in ordered_operators:
+        for symbol in op.user_symbols:
```

Comment:
I feel like instead of looping through all the operators for every term, we should just lay out a mapping from user-inputted tokens to the objects that represent that operator. We could probably get rid of the `user_symbols` class variable at that point too and just keep one `canonical_symbol` that is used in `to_sql`